# Advanced Certification Program in Computational Data Science
## A program by IISc and TalentSprint
### Mini Project Notebook: Linear Algebra and Calculus

## Problem Statement

 The task is to advise a petroleum company on how to meet the demands of their customers for motor oil, diesel oil and gasoline.

## Learning Objectives

At the end of the experiment, you will be able to

* create arrays and matrices in python
* understand the concepts of linear equations
* solve the system of linear equations

### Data

From a barrel of crude oil, in one day, factory $A$ can produce
* 20 gallons of motor oil,
* 10 gallons of diesel oil, and
* 5 gallons of gasoline

Similarly, factory $B$ can produce
* 4 gallons of motor oil,
* 14 gallons of diesel oil, and
* 5 gallons of gasoline

while factory $C$ can produce
* 4 gallons of motor oil,
* 5 gallons of diesel oil, and
* 12 gallons of gasoline

There is also waste in the form of paraffin, among other things. Factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude, factory $B$ 5 gallons, and factory $C$ 2 gallons.

**Note:** Your conclusion should include a discussion of the nature of the terms *unique*, *no solution*, *overdetermined* and *underdetermined* as they apply in the context of the oil plants.

## Grading = 10 Points

### Create an array

Create an array of size 2x3 with arbitrary values.

In [ ]:
# YOUR CODE HERE
import numpy as np
#array = np.random.random((2,3))
array = np.array([[-4,2,6], [1,-5,3]])
array

array([[-4,  2,  6],
       [ 1, -5,  3]])

### Create the system of Linear Equations

Suppose the current daily demand from distributors is 6600 gallons of motor oil, 5100 gallons of diesel oil and 3100 of gasoline.

Set up the system of equations which describes the above situation. Please include the units as well.

Let the number of barrels used by factory $A$, $B$ and $C$ are $x$, $y$ and $z$ respectively.

Then the system of linear equations will be

$$Motor\ oil:\ \ \ 20x + 4y + 4z = 6600$$

$$Diesel\ oil:\ \ \ 10x + 14y + 5z = 5100$$

$$Gasoline:\ \ \ 5x + 5y + 12z = 3100$$

### Solve the system of Linear Equation (2 points)

How many barrels of crude oil each plant should get in order to meet the demand as a group. Remember that we can only provide each plant with an integral number of barrels.

In [ ]:
# YOUR CODE HERE
A = np.array([[20, 4, 4], [10, 14, 5], [5, 5, 12]])
b1 = np.array([6600, 5100,  3100])

def check_unique_solution(A, b):
  augmented_matrix = np.column_stack((A,b1))
  rank_A = np.linalg.matrix_rank(A)
  rank_augmented = np.linalg.matrix_rank(augmented_matrix)
  if rank_A == rank_augmented and rank_A == A.shape[1]:
    print("The system has a unique solution.")
  else:
    print("The system does not have a unique solution")

check_unique_solution(A,b1)
x1 = np.ceil(np.linalg.solve(A,b1))
print(f'The number of barrels required by factory A, B and C to meet the demand are {x1[0]}, {x1[1]} and {x1[2]} respectively')

The system has a unique solution.
The number of barrels required by factory A, B and C to meet the demand are 288.0, 129.0 and 85.0 respectively


In [ ]:
# Double Check if demand is met
np.dot(A, x1) >= b1

array([ True,  True,  True])

Suppose the total demand for all products **doubled**. What would the solution now be? How does it compare to the original solution? Why, mathematically, should this have been expected?

In [ ]:
# YOUR CODE HERE
b2 = 2*b1
check_unique_solution(A,b2)
x2 = np.ceil(np.linalg.solve(A, b2))
print(f'The number of barrels required by factory A, B and C to meet the doubled demand are {x2[0]}, {x2[1]} and {x2[2]} respectively')

The system has a unique solution.
The number of barrels required by factory A, B and C to meet the doubled demand are 575.0, 258.0 and 170.0 respectively


In [ ]:
np.linalg.solve(A, b2)/np.linalg.solve(A,b1)

array([2., 2., 2.])

The number of barrels got doubled by doubling the demand. We can represent x1 = A^-1 * b1, x2 = A^-1 * (2 b1) = 2 (A^-1 * b1) = 2 * x1


Suppose that the company acquires another group of distributors and that the daily demand of this group is 2000 gallons of motor oil, 4000 gallons of gasoline, and 4000 gallons of diesel oil. How would you set up production of just this supply? Are there any options (more than one way)?

In [ ]:
# YOUR CODE HERE
b3 = np.array([2000, 4000, 4000])
check_unique_solution(A,b3)
x3 = np.ceil(np.linalg.solve(A, b3))
print(f'The number of barrels required by factory A, B and C to meet the new demand are {x3[0]}, {x3[1]} and {x3[2]} respectively')

The system has a unique solution.
The number of barrels required by factory A, B and C to meet the new demand are 13.0, 188.0 and 250.0 respectively


In [ ]:
import sympy as sy
augmented_matrix = sy.Matrix(np.column_stack((A,b3)))
augmented_matrix.rref()

(Matrix([
 [1, 0, 0,  25/2],
 [0, 1, 0, 375/2],
 [0, 0, 1,   250]]),
 (0, 1, 2))

We have a unique solution to the above augmented matrix. Hence, there's only one way.

Next, calculate the needs of each factory (in barrels of crude, as usual) to meet the total demand of both groups of distributors. When you have done this, compare your answer to results already obtained. What mathematical conclusion can you draw?

In [ ]:
# YOUR CODE HERE
b4 = b1 + b3
check_unique_solution(A,b4)
x4 = np.ceil(np.linalg.solve(A, b4))
print(f'The number of barrels required by factory A, B and C to meet the total demand of both groups of distributors are {x4[0]}, {x4[1]} and {x4[2]} respectively')

The system has a unique solution.
The number of barrels required by factory A, B and C to meet the total demand of both groups of distributors are 300.0, 317.0 and 335.0 respectively


In [ ]:
np.linalg.solve(A, b4) == np.linalg.solve(A, b1) + np.linalg.solve(A, b3)

array([ True,  True,  True])

x4 = A^-1 * b4 = A^-1 * (b1 + b3) = (A^-1 * b1) + (A^-1 * b3) = x1 + x3

### Sensitivity and Robustness (1 point)

In real life applications, constants are rarely ever exactly equal to their stated value; certain amounts of uncertainty are always present. This is part of the reason for the science of statistics. In the above model, the daily productions for the plants would be averages over a period of time. Explore what effect small changes in the parameters have on the output.

To do this, pick any 3 coefficients, one at a time, and increase or decrease them by 3%. For each case , note what effect this has on the solution, as a percentage change. Can you draw any overall conclusion?

In [ ]:
# YOUR CODE HERE
A_perturbed = np.hstack(((A[:,0] * 1.03).reshape(-1,1), A[:,1:]))
check_unique_solution(A_perturbed,b1)
(np.linalg.solve(A_perturbed, b1) - np.linalg.solve(A, b1)) * 100 / np.linalg.solve(A, b1)

The system has a unique solution.


array([-2.91262136,  0.        ,  0.        ])

Increasing the coefficient of factory A by 3% has decreased the barrel used by factory A by around 3% and that of factory B and C is unchanged.

In [ ]:
A_perturbed = np.hstack(((A[:,0] * 0.97).reshape(-1,1), A[:,1:]))
check_unique_solution(A_perturbed,b1)
(np.linalg.solve(A_perturbed, b1) - np.linalg.solve(A, b1)) * 100 / np.linalg.solve(A, b1)

The system has a unique solution.


array([3.09278351, 0.        , 0.        ])

Decreasing the coefficient of factory A by 3% has increased the barrel used by factory A by around 3% and that of factory B and C is unchanged.

In [ ]:
A_perturbed = np.hstack((A[:,0].reshape(-1,1) , (A[:,1] * 1.03).reshape(-1,1), A[:,2].reshape(-1,1)))
check_unique_solution(A_perturbed,b1)
(np.linalg.solve(A_perturbed, b1) - np.linalg.solve(A, b1)) * 100 / np.linalg.solve(A, b1)

The system has a unique solution.


array([ 0.00000000e+00, -2.91262136e+00, -1.67186526e-14])

In [ ]:
A_perturbed = np.hstack((A[:,0].reshape(-1,1) , (A[:,1] * 0.97).reshape(-1,1), A[:,2].reshape(-1,1)))
check_unique_solution(A_perturbed,b1)
(np.linalg.solve(A_perturbed, b1) - np.linalg.solve(A, b1)) * 100 / np.linalg.solve(A, b1)

The system has a unique solution.


array([0.        , 3.09278351, 0.        ])

In [ ]:
A_perturbed = np.hstack((A[:,:2], (A[:,2] * 1.03).reshape(-1,1)))
check_unique_solution(A_perturbed,b1)
(np.linalg.solve(A_perturbed, b1) - np.linalg.solve(A, b1)) * 100 / np.linalg.solve(A, b1)

The system has a unique solution.


array([ 0.        ,  0.        , -2.91262136])

In [ ]:
A_perturbed = np.hstack((A[:,:2], (A[:,2] * 0.97).reshape(-1,1)))
check_unique_solution(A_perturbed,b1)
(np.linalg.solve(A_perturbed, b1) - np.linalg.solve(A, b1)) * 100 / np.linalg.solve(A, b1)

The system has a unique solution.


array([0.        , 0.        , 3.09278351])

Similar behavior with factory B and C coefficient.

Overall conclusion : The system is robust and stable. Increasing the coefficient value by x% decreases the barrel used by that factory by x% approximately and vice-versa.

### A Plant Off-Line (1 point)

Suppose factory $C$ is shut down by the EPA (Environmental Protection Agency) temporarily for excessive emissions into the atmosphere. If your demand is as it was originally (6600, 5100, 3100), what would you now say about the companies ability to meet it? What do you recommend they schedule for production now?

20x + 4y = 6600


10x + 14y = 5100


5x + 5y = 3100


3 equations and 2 unknows. Over-determined system.

In [ ]:
# YOUR CODE HERE
A_new = A[:,:2]
A_new

array([[20,  4],
       [10, 14],
       [ 5,  5]])

In [ ]:
check_unique_solution(A_new,b1)

The system does not have a unique solution


In [ ]:
# Normal equation approach
print('Rank of normal equation : ',np.linalg.matrix_rank(np.dot(A_new.T, A_new)))
solve_normal = np.ceil(np.linalg.solve(np.dot(A_new.T, A_new), np.dot(A_new.T, b1)))
solve_normal

Rank of normal equation :  2


array([300., 169.])

In [ ]:
np.dot(A_new, solve_normal) >= b1

array([ True,  True, False])

In [ ]:
solve12 = np.ceil(np.linalg.solve(A_new[:2,:], b1[:2])) # Use 1st and 2nd equations
solve23 = np.ceil(np.linalg.solve(A_new[1:,:], b1[1:])) # Use 2nd and 3rd equations
solve13 = np.ceil(np.linalg.solve(A_new[[0,2],:], b1[[0,2]])) # Use 1st and 3rd equations
solve12, solve23, solve13

(array([300., 150.]), array([ 895., -275.]), array([258., 363.]))

In [ ]:
np.dot(A_new, solve12) == b1

array([ True,  True, False])

In [ ]:
np.dot(A_new, solve13) == b1

array([False, False, False])

In [ ]:
np.dot(A_new, solve13) >= b1

array([ True,  True,  True])

In [ ]:
solve13

array([258., 363.])

The solution of normal equation does not satisfy the minimum gosoline requirment.
The solution with last two equations dont make sense as the value cant be negative. The solution with 1st and 2nd equation doesn't satisfy the requirment for gasoline. The solution with 1st and 3rd equation doesn't satisfy equation 2 exactly. There is no unique solution satisfying all equations simultaneously. However, the solution with 1st and 3rd equation satifies the minimum requirement demand. Would recommend x=258., y=363. barrels as the schedule for production.

### Buying another plant

####(Note the following given information. You will see questions in continuation to this, in the subsequent sections)

This situation has caused enough concern that the CEO is considering buying another plant, identical to the third, and using it permanently. Assuming that all 4 plants are on line, what production do you recommend to meet the current demand (5000, 8500, 10000)? In general, what can you say about any increased flexibility that the 4th plant might provide?

Let the number of barrels used by factory $A$, $B$, $C$ and $D$ are $x$, $y$, $z$ and $w$ respectively.

Then the system of linear equations will be

$$20x + 4y + 4z + 4w = 5000$$

$$10x + 14y + 5z + 5w = 8500$$

$$5x + 5y + 12z + 12w = 10000$$

The above system of linear equation has fewer equations than variables, hence it is *underdetermined* and cannot have a unique solution. In this case, there are either infinitely many solutions or no exact solution. We can solve it by keeping $w$ as constant and using [rref](http://linear.ups.edu/html/section-RREF.html) form to solve the system of linear equation.

To know about rref implementation in python refer [here](https://docs.sympy.org/latest/tutorial/matrices.html#rref).

In [ ]:
import sympy as sy

# create symbol 'w'
w = sy.Symbol("w")
A_aug = sy.Matrix([[20, 4, 4, 5000-4*w],
                   [10, 14, 5, 8500-5*w],
                   [5, 5, 12, 10000-12*w]])
# show rref form
mat, _ = A_aug.rref()
print(mat)
print(f'The number of barrels required by factory A and B to meet the demand of distributors are {np.ceil(mat[0,3])}, {np.ceil(mat[1,3])} respectively, C + D = 675 barrels')

Matrix([[1, 0, 0, 195/4], [0, 1, 0, 1325/4], [0, 0, 1, 675 - w]])
The number of barrels required by factory A and B to meet the demand of distributors are 49, 332 respectively, C + D = 675 barrels


x = 195/4 , y = 1325/4, z = 675-w. We have infinite solutions [w and z should be chosen in such a way that w+z = 675].

From the above result, it can be seen that 4th plant will share the number of barrels required by the 3rd plant only, while the requirement of 1st and 2nd plant will remain unaffected.

### Calculate the amount of Paraffin supplied (1 point)

The company has just found a candle company that will buy its paraffin. Under the current conditions (i.e, after buying another plant) for demand (5000, 8500, 10000), how much can be supplied to them per day?

According to the problem statement, factory $A$ has 3 gallons of paraffin to dispose of per barrel of crude oil, factory $B$ 5 gallons, and factory $C$ 2 gallons.

In [ ]:
# YOUR CODE HERE
# Assume 2 gallons per barrel of crude oil from factory D as well
paraffin_A = np.ceil(195.0/4) * 3
paraffin_B = np.ceil(1325.0/4) * 5
paraffin_C_D = np.ceil(675) * 2
print(f'Amount of paraffin that can be supplied per day is {paraffin_A + paraffin_B + paraffin_C_D} gallons')

Amount of paraffin that can be supplied per day is 3157.0 gallons


### Selling the first plant (1 point)

The management is also considering selling the first plant due to aging equipment and high workman's compensation costs for the state it is located in. They would like to know what this would do to their production capability. Specifically, they would like an example of a demand they could not meet with only plants 2 and 3, and also what effect having plant 4 has (recall it is identical to plant 3). They would also like an example of a demand that they could meet with just plants 2 and 3. Any general statements you could make here would be helpful.

Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

When considering only plants 2 and 3, and demand (5000, 8500, 10000) then we have

$$4y + 4z = 5000$$

$$14y + 5z = 8500$$

$$5y + 12z = 10000$$

3 equations and 2 unknows. Overdetermined system.

In [ ]:
# YOUR CODE HERE
A_new = np.array([[4,4], [14,5], [5,12]])
demand = np.array([5000, 8500, 10000])
check_unique_solution(A_new, demand)
print('Rank of normal equation : ', np.linalg.matrix_rank(np.dot(A_new.T, A_new)))
solve_normal = np.linalg.solve(np.dot(A_new.T, A_new), np.dot(A_new.T, demand))
solve12 = np.linalg.solve(np.array([[4,4], [14,5]]), np.array([5000, 8500])) # 1st and 2nd equation
solve13 = np.linalg.solve(np.array([[4,4], [5,12]]), np.array([5000, 10000])) # 1st and 3rd equation
solve23 = np.linalg.solve(np.array([[14,5], [5,12]]), np.array([8500, 10000])) # 2nd and 3rd equation
np.ceil(solve12), np.ceil(solve13), np.ceil(solve23), np.ceil(solve_normal)

The system does not have a unique solution
Rank of normal equation :  2


(array([ 250., 1000.]),
 array([715., 536.]),
 array([364., 682.]),
 array([370., 696.]))

In [ ]:
np.dot(A_new, solve12) >= demand

array([ True,  True,  True])

In [ ]:
np.dot(A_new, solve13) >= demand

array([ True,  True,  True])

In [ ]:
np.dot(A_new, solve23) >= demand

array([False,  True,  True])

In [ ]:
np.dot(A_new, solve_normal) >= demand

array([False,  True,  True])

No unique solution satisfying all equations simulateously. But if we are looking for meeting the minimum demand, then solution using solving 1st and 2nd equation or 1st and 3rd equation would solve our purpose.

Taking 4th plant into consideration.
Let the number of barrels used by factory $B$, $C$ and $D$ are $y$, $z$ and $w$ respectively.

Then for demand (5000, 8500, 10000) the system of linear equations will be

$$4y + 4z + 4w = 5000$$

$$14y + 5z + 5w = 8500$$

$$5y + 12z + 12w = 10000$$

Solve it using rref form.

In [ ]:
# YOUR CODE HERE
check_unique_solution(np.matrix([[4,4,4], [14,5,5], [5,12,12]]), np.matrix([5000, 8500, 10000]))
Augmented_matrix = sy.Matrix([[4, 4, 4, 5000], [14, 5, 5, 8500], [5, 12, 12, 10000]])
Augmented_matrix.rref()

The system does not have a unique solution


(Matrix([
 [1, 0, 0, 0],
 [0, 1, 1, 0],
 [0, 0, 0, 1]]),
 (0, 1, 3))

In [ ]:
BCD = np.matrix([[4,4,4], [14,5,5], [5,12,12]])
np.linalg.matrix_rank(BCD)

2

No exact solution. z+w = 0. But z or w can't be negative. Only possible way is z=w=0. It's equivalent to shutting down plant C and D.

Now, changing demand to (6600, 5100, 3100) and solving the system of equation using rref form.

In [ ]:
# YOUR CODE HERE
check_unique_solution(np.matrix([[4,4,4], [14,5,5], [5,12,12]]), np.matrix([6600, 5100, 3100]))
Augmented_matrix = sy.Matrix([[4, 4, 4, 6600], [14, 5, 5, 5100], [5, 12, 12, 3100]])
Augmented_matrix.rref()

The system does not have a unique solution


(Matrix([
 [1, 0, 0, 0],
 [0, 1, 1, 0],
 [0, 0, 0, 1]]),
 (0, 1, 3))

No exact solution. z+w = 0. But z or w can't be negative. Only possible way is z=w=0. It's equivalent to shutting down plant C and D. Irrespective of the demand, the system has no exact solution as system is rank deficient.

### Set rates for Products (1 point)

Company wants to set the rates of motor oil, diesel oil, and gasoline. For this purpose they have few suggestions given as follows:

* 100, 66, 102 Rupees per gallon,

* 104, 64, 100 Rupees per gallon,

* 102, 68, 98 Rupees per gallon, and

* 96, 68, 100 Rupees per gallon

for motor oil, diesel oil, and gasoline respectively.

Using matrix multiplication, find the rates which result in maximum total price.

Let $M$ denote the matrix such that rows represents different plants (A, B and C), columns represents different products (motor oil, diesel oil and gasoline) and each value represents production of that product from one barrel of crude oil for that plant.

$$M = \begin{bmatrix}
20 & 10 & 5 \\
4 & 14 & 5  \\
 4 & 5 & 12  
\end{bmatrix}$$

Also, $R$ is a matrix having different rates as its columns.

$$R = \begin{bmatrix}
100 & 104 & 102 & 96 \\
66 & 64 & 68 & 68  \\
102 & 100 & 98 & 100  
\end{bmatrix}$$

In [ ]:
# YOUR CODE HERE
M = np.array([[20, 10, 5], [4, 14, 5], [4, 5, 12]])
R = np.array([[100, 104, 102, 96], [66, 64, 68, 68], [102, 100, 98, 100]])
price = np.dot(M, R)
price

array([[3170, 3220, 3210, 3100],
       [1834, 1812, 1850, 1836],
       [1954, 1936, 1924, 1924]])

In [ ]:
total_price = sum(price)
max_index = np.argmax(total_price)
rate = R[:, max_index]
rate

array([102,  68,  98])

Suggested rate in rupees per gallon for motor oil, diesel oil, and gasoline respectively to get maximum total price is [102, 68, 98]

### Marginal Cost (1 point)

The total cost $C(x)$ in Rupees, associated with the production of $x$ gallons of gasoline is given by

$$C(x) = 0.005 x^3 – 0.02 x^2 + 30x + 5000$$

Find the marginal cost when $22$ gallons are produced, where, marginal cost means the instantaneous rate of change of total cost at any level of output.

In [ ]:
# YOUR CODE HERE
import sympy as sp
x = sp.Symbol("x")
C = 0.005 * x**3 - 0.02 * x**2 + 30 * x + 5000
C_prime = sp.diff(C, x)
marginal_cost = C_prime.subs(x, 22)
marginal_cost

36.3800000000000

### Marginal Revenue (1 point)

The total revenue in Rupees received from the sale of $x$ gallons of a motor oil is given by $$R(x) = 3x^2 + 36x + 5.$$

Find the marginal revenue, when $x = 28$, where, marginal revenue means the rate of change of total revenue with respect to the number of items sold at an instant.

In [ ]:
# YOUR CODE HERE
R = 3 * x**2 + 36 * x + 5
R_prime = sp.diff(R, x)
marginal_revenue = R_prime.subs(x, 28)
print(marginal_revenue)

204


### Pouring crude oil in tank (1 point)

In a cylindrical tank of radius 10 meter, crude oil is being poured at the rate of 314 cubic meter per hour. Then find

* the rate at which the height of crude oil is increasing in the tank, and
* the height of crude oil in tank after 2 hours.

V = pi*  r * r * h

dV/dt = pi * r * r * dh/dt

In [ ]:
# YOUR CODE HERE
from sympy import symbols, diff
r = 10
h = symbols('h')
V = np.pi * r**2 * h
dVdt = diff(V,h) * symbols('dh/dt')  # chain rule
dh_dt = (314/diff(V,h)).evalf()
print('Rate at which height of crude oil is increasing in the tank: ', dh_dt)

Rate at which height of crude oil is increasing in the tank:  0.999493042617103


The rate at which the height of crude oil is increasing in the tank is 0.9995 meter per hour.

In [ ]:
from sympy import symbols, integrate
t = symbols('t')
h = integrate(dh_dt, (t,0,2))

#t=2
#h = dhdt * t
print('Height of crude oil in tank after 2 hours: ', h)


Height of crude oil in tank after 2 hours:  1.99898608523421


The height of crude oil in tank after 2 hours is 1.999 meter.